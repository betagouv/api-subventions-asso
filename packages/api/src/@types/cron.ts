export interface CronController {
    name: string;
}
