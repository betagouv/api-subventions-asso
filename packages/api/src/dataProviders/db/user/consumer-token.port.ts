import { ObjectId } from "mongodb";
import { ConsumerToken } from "../../../modules/user/entities/ConsumerToken";
import MongoPort from "../../../shared/MongoPort";

class ConsumerTokenPort extends MongoPort<ConsumerToken> {
    collectionName = "consumer-token";

    async findToken(userId: string | ObjectId) {
        return (await this.collection.findOne({ userId: new ObjectId(userId) }))?.token;
    }

    async find(userId: string) {
        return this.collection.find({ userId: new ObjectId(userId) }).toArray();
    }

    async create(entity: ConsumerToken) {
        return (await this.collection.insertOne(entity)).acknowledged;
    }

    async deleteAllByUserId(userId: string | ObjectId) {
        const result = await this.collection.deleteMany({ userId: new ObjectId(userId) });
        return result.acknowledged;
    }

    async createIndexes() {
        await this.collection.createIndex({ userId: 1 });
    }
}

const consumerTokenPort = new ConsumerTokenPort();

export default consumerTokenPort;
