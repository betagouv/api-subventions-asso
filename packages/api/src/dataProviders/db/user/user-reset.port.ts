import { ObjectId } from "mongodb";
import UserReset from "../../../modules/user/entities/UserReset";
import MongoRepository from "../../../shared/MongoRepository";

export class UserResetPort extends MongoRepository<UserReset> {
    collectionName = "users-reset";

    public async findByToken(token: string) {
        return this.collection.findOne({ token });
    }

    public async findByUserId(userId: ObjectId | string) {
        return this.collection.find({ userId: new ObjectId(userId) }).toArray();
    }

    public async findOneByUserId(userId: ObjectId | string) {
        return this.collection.findOne({ userId: new ObjectId(userId) });
    }

    public async create(reset: UserReset) {
        await this.collection.insertOne(reset);
        return reset;
    }

    public async remove(reset: UserReset) {
        return this.collection.deleteOne(reset);
    }

    public async removeAllByUserId(userId: ObjectId) {
        const result = await this.collection.deleteMany({ userId });
        return result.acknowledged;
    }

    async createIndexes() {
        await this.collection.createIndex({ token: 1 }, { unique: true });
        await this.collection.createIndex({ userId: 1 }, { unique: true });
    }
}

const userResetPort = new UserResetPort();

export default userResetPort;
