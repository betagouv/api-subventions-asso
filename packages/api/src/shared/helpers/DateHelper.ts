import { capitalizeFirstLetter } from "./StringHelper";

export const isDateNewer = (a: string | Date, b: string | Date) => {
    const dateA = new Date(a);
    const dateB = new Date(b);
    if (!dateA || !dateB) return false;
    const diff = dateA.getTime() - dateB.getTime();
    if (diff >= 0) return true;
    else return false;
};

export const frenchToEnglishMonthsMap = {
    JANVIER: "january",
    FEVRIER: "february",
    MARS: "march",
    AVRIL: "april",
    MAI: "may",
    JUIN: "june",
    JUILLET: "july",
    AOUT: "august",
    SEPTEMBRE: "september",
    OCTOBRE: "october",
    NOVEMBRE: "november",
    DECEMBRE: "december"
};

export const englishMonthNames = Object.values(frenchToEnglishMonthsMap).map(monthLowercase =>
    capitalizeFirstLetter(monthLowercase)
);

export const getMonthFromFrenchStr = (month: string) => {
    return frenchToEnglishMonthsMap[month];
};

export const isValidDate = date => date instanceof Date && !isNaN(date as unknown as number);
