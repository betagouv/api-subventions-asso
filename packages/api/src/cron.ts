import { ToadScheduler } from "toad-scheduler";
import { CronController } from "./@types/cron";
// eslint-disable-next-line @typescript-eslint/no-unused-vars -- test purposes
import { ExampleCronController } from "./modules/example/interfaces/cron/example.cron.controller";
import { DemarchesSimplifieesCronController } from "./modules/example/interfaces/cron/demarchesSimplifiees.cron.controller";
import { DauphinCronController } from "./modules/providers/dauphin/interfaces/cron/dauphin.cron.controller";
import { DumpCronController } from "./modules/dump/interfaces/cron/DumpCronController";

const controllers: CronController[] = [
    // ExampleCronController,
    DemarchesSimplifieesCronController,
    DauphinCronController,
    DumpCronController,
];

export const scheduler = new ToadScheduler();

export function initCron() {
    for (const ControllerClass of controllers) {
        // @ts-expect-error generic type
        const controller = new ControllerClass();
        for (const job of controller["__intervalJobs__"] || []) {
            scheduler.addIntervalJob(job);
        }
        for (const job of controller["__cronJobs__"] || []) {
            scheduler.addCronJob(job);
        }
    }
}
