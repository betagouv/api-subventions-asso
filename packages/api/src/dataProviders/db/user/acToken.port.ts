import { ObjectId } from "mongodb";
import MongoRepository from "../../../shared/MongoRepository";
import { AgentConnectTokenDbo } from "../../../modules/user/@types/AgentConnectUser";

class AgentConnectTokenRepository extends MongoRepository<AgentConnectTokenDbo> {
    collectionName = "agent-connect-token";

    findLastActive(userId: ObjectId) {
        return this.collection.findOne({ userId: new ObjectId(userId) }, { sort: { creationDate: -1 } });
    }

    async upsert(entity: Omit<AgentConnectTokenDbo, "_id">) {
        return (await this.collection.updateOne({ userId: entity.userId }, { $set: entity }, { upsert: true }))
            .acknowledged;
    }

    async deleteAllByUserId(userId: string | ObjectId) {
        const result = await this.collection.deleteMany({ userId: new ObjectId(userId) });
        return result.acknowledged;
    }

    async createIndexes() {
        await this.collection.createIndex({ userId: 1 }, { unique: true });
        await this.collection.createIndex({ expiresAt: 1 });
    }
}

const agentConnectTokenRepository = new AgentConnectTokenRepository();
export default agentConnectTokenRepository;
