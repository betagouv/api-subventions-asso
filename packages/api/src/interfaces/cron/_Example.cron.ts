import { AsyncCron, AsyncIntervalCron, Cron, IntervalCron } from "../../decorators/cronController.decorator";
import { CronController } from "../../@types/cron";

export class ExampleCron implements CronController {
    name = "cronExample";
    static testParity = 0;

    @IntervalCron({ minutes: 1 }, false)
    intervalTest() {
        console.log("hello world");
    }

    @AsyncIntervalCron({ minutes: 1 }, false)
    asyncIntervalTest() {
        ExampleCron.testParity += 1;
        if (ExampleCron.testParity % 2) return Promise.reject("odd number: task fails");
        console.log("even number: test successes");
        return Promise.resolve();
    }

    @Cron({ cronExpression: "* * * * *" })
    test() {
        console.log("good morning world");
    }

    @AsyncCron({ cronExpression: "* * * * *" })
    asyncTest() {
        ExampleCron.testParity += 1;
        if (ExampleCron.testParity % 2) return Promise.reject("odd number: cron task fails");
        console.log("even number: test successes");
        return Promise.resolve();
    }
}
