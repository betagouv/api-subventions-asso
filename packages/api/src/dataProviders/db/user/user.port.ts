import { UserDto } from "dto";
import { Filter, ObjectId } from "mongodb";
import { buildDuplicateIndexError, isMongoDuplicateError } from "../../../shared/helpers/MongoHelper";
import MongoPort from "../../../shared/MongoPort";
import { removeHashPassword, removeSecrets } from "../../../shared/helpers/PortHelper";
import { InternalServerError } from "../../../shared/errors/httpErrors";
import UserDbo, { UserNotPersisted } from "./UserDbo";

export class UserPort extends MongoPort<UserDbo> {
    collectionName = "users";

    joinIndexes = {
        associationVisits: "_id",
    };

    async findAll() {
        return this.collection.find({}).toArray();
    }

    async findByEmail(email: string) {
        const user = await this.collection.findOne({ email: email });
        if (!user) return null;
        return removeSecrets(user);
    }

    async find(query: Filter<UserDbo> = {}): Promise<UserDto[]> {
        const dbos = await this.collection.find(query).toArray();
        return dbos.map(dbo => removeSecrets(dbo));
    }

    async findById(userId: ObjectId | string): Promise<UserDto | null> {
        const user = await this.collection.findOne({ _id: new ObjectId(userId) });
        if (!user) return null;
        return removeSecrets(user);
    }

    async findByPeriod(begin: Date, end: Date, withAdmin): Promise<UserDto[]> {
        const query: Filter<UserDbo> = { signupAt: { $gte: begin, $lt: end } };
        if (!withAdmin) query.roles = { $ne: "admin" };
        return this.find(query);
    }

    async findInactiveSince(date: Date): Promise<UserDto[]> {
        const query: Filter<UserDbo> = {
            lastActivityDate: { $lt: date },
            roles: { $ne: "admin" },
            disable: { $ne: true },
        };
        return this.find(query);
    }

    async findNotActivatedSince(date: Date, lastWarned: Date | undefined = undefined): Promise<UserDto[]> {
        const query: Filter<UserDbo> = {
            lastActivityDate: null,
            signupAt: { $lt: date },
            roles: { $ne: "admin" },
            disable: { $ne: true },
        };
        // @ts-expect-error -- query too strictly typed
        if (lastWarned) query.signupAt["$gt"] = lastWarned;
        return this.find(query);
    }

    async update(user: Partial<UserDbo>, withJwt = false): Promise<UserDto | Omit<UserDbo, "hashPassword">> {
        const res = user._id
            ? await this.collection.findOneAndUpdate({ _id: user._id }, { $set: user }, { returnDocument: "after" })
            : await this.collection.findOneAndUpdate(
                  { email: user.email },
                  { $set: user },
                  { returnDocument: "after" },
              );

        if (!res) throw new InternalServerError("User update failed");
        return withJwt ? removeHashPassword(res) : removeSecrets(res);
    }

    async delete(user: UserDto): Promise<boolean> {
        const result = await this.collection.deleteOne({ _id: user._id });
        return result.acknowledged;
    }

    async create(user: UserNotPersisted): Promise<UserDto> {
        return removeSecrets(await this.createAndReturnWithJWT(user));
    }

    async createAndReturnWithJWT(user: UserNotPersisted): Promise<Omit<UserDbo, "hashPassword">> {
        const userDbo: UserDbo = { ...user, _id: new ObjectId() };

        try {
            await this.collection.insertOne(userDbo);
        } catch (error) {
            if (isMongoDuplicateError(error)) throw buildDuplicateIndexError<UserDbo>(error);
            throw error;
        }
        return removeHashPassword(userDbo);
    }

    async getUserWithSecretsByEmail(email: string): Promise<UserDbo | null> {
        return this.collection.findOne({ email });
    }

    async getUserWithSecretsById(id: ObjectId): Promise<UserDbo | null> {
        return this.collection.findOne({ _id: id });
    }

    countTotalUsersOnDate(date, withAdmin: boolean) {
        const query: Filter<UserDbo> = { signupAt: { $lt: date } };
        if (!withAdmin) query.roles = { $ne: "admin" };
        return this.collection.find(query).count();
    }

    async createIndexes() {
        await this.collection.createIndex({ email: 1 }, { unique: true });
    }
}

const userPort = new UserPort();

export default userPort;
