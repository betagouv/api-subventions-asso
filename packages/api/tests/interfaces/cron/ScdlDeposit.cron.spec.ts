import depositLogPort from "../../../src/dataProviders/db/deposit-log/depositLog.port";
import userPort from "../../../src/dataProviders/db/user/user.port";
import { ScdlDepositCron } from "../../../src/interfaces/cron/ScdlDeposit.cron";
import { DEPOSIT_LOG_DBO } from "../../../src/modules/deposit-scdl-process/__fixtures__/depositLog.fixture";
import brevoMailNotifyPipe from "../../../src/modules/notify/outPipes/BrevoMailNotifyPipe";
import { USER_NOT_PERSISTED } from "../../../src/modules/user/__fixtures__/user.fixture";
import { addDaysToDate } from "../../../src/shared/helpers/DateHelper";

describe("ScdlDeposit CRON", () => {
    let cron: ScdlDepositCron;

    beforeEach(() => {
        cron = new ScdlDepositCron();
    });

    describe("notifyUsers", () => {
        beforeEach(async () => {});

        it("send mail to users that started deposit 2 days ago", async () => {
            const now = new Date();

            const user = await userPort.create({
                ...USER_NOT_PERSISTED,
                signupAt: addDaysToDate(now, -10),
            });
            jest.useFakeTimers().setSystemTime(addDaysToDate(now, -2)); // mock updateDate from deposit date creation
            await depositLogPort.insertOne({
                ...DEPOSIT_LOG_DBO,
                userId: user._id.toString(),
            });
            jest.useRealTimers();
            await cron.notifyUsers();
            // @ts-expect-error: mocked
            const args = jest.mocked(brevoMailNotifyPipe.apiInstance.sendTransacEmail).mock.calls[0][0];
            expect(args).toMatchSnapshot();
        });
    });
});
