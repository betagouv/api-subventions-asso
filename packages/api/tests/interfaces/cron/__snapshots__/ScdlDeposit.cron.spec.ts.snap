// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScdlDeposit CRON notifyUsers send mail to users that started deposit 2 days ago 1`] = `
{
  "bcc": [
    {
      "email": "log@datasubvention.beta.gouv.fr",
      "name": "Data.Subvention Log",
    },
  ],
  "params": {},
  "sender": {
    "email": "mail-user@datasubvention",
    "name": "Data.Subvention",
  },
  "templateId": 262,
  "to": [
    {
      "email": "user@beta.gouv.fr",
    },
  ],
}
`;
